Opencv v3.4.3

hmaarrfk opened the pull request:
Checklist
* [x] Used a fork of the feedstock to propose changes
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/conda_smithy.html#how-to-re-render ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.


I used multiple source files for the contrib so that so maybe conda-smithy can auto update the hash in the future? Anyway, should help iterating on builds.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found it was in an excellent condition.

hmaarrfk commented:
Is it OK to add my patch for bayer_XX2rgba 

https://github.com/opencv/opencv/pull/12465

It would mean alot to me!

hmaarrfk commented:
@conda-forge/opencv do you know why on windows it is finding python 3.4?

https://ci.appveyor.com/project/conda-forge/opencv-feedstock/build/1.0.268/job/wr38yf2kiock0vh6?fullLog=true#L1917

when just above it it claims it is version 3.6
https://ci.appveyor.com/project/conda-forge/opencv-feedstock/build/1.0.268/job/wr38yf2kiock0vh6?fullLog=true#L1784

hajapy commented:
I don't know why the windows build is failing, though I noted the same problem here: https://github.com/conda-forge/opencv-feedstock/pull/120

I didn't have any time try debugging it..

hmaarrfk commented:
Maybe @jakirkham can give us some of his wisdom from 2 years back in what looked like a colossal effort to get 3.1.0 working

https://github.com/conda-forge/opencv-feedstock/pull/16

https://github.com/conda-forge/opencv-feedstock/commit/3f7eaed19e6b7d58b2ba5bb536a154fd266670e1#diff-d04c86b6bb20341f5f7c53165501a393

hmaarrfk commented:
Now only mac is failing. If you have any clue on why it would be greatly appreciated!

jakirkham commented on recipe/meta.yaml:
```diff
@@ -1,36 +1,37 @@
-    - curl >=7.44.0,<8
     - cmake
     - {{ compiler('c') }}
     - {{ compiler('cxx') }}
```
Looks like we need `pkg-config` here.

ref: https://travis-ci.org/conda-forge/opencv-feedstock/jobs/426194306#L815

jakirkham commented on recipe/meta.yaml:
```diff
@@ -1,36 +1,37 @@
     - {{ compiler('cxx') }}
   host:
     - python
     - pkgconfig                    # [unix]
```
Admittedly this has been here for a long time, but this is probably unneeded and should be dropped.

Unfortunately this is a little confusing, `pkgconfig` is a Python module that allows one to use `pkg-config` (the Unix build tool). AFAICT OpenCV doesn't use the Python package, but it does rely on the `pkg-config` command line tool and `.pc` files, which are expected to be in the build prefix (likely the reason FFMPEG is not being picked up).

hmaarrfk replied:
This was removed in @hajapy 's 3.4.2 PR.

Thanks for the explanation on pkgconfig

jakirkham replied:
Sure. For clarity, `pkgconfig` should be dropped from `host` as well.

jakirkham commented on recipe/run_test.py:
```diff
@@ -7,7 +7,7 @@
 
 
-OPENCV_AVI_URL = 'https://github.com/opencv/opencv_extra/raw/master/testdata/highgui/video/VID00003-20100701-2204.avi'
+OPENCV_AVI_URL = 'https://github.com/opencv/opencv_extra/raw/master/testdata/highgui/video/VID00003-20100701-2204.mpg'
```
This can probably be reverted as it doesn't appear related to the issue.

jakirkham commented:
Thanks for working on this @hmaarrfk. Very much appreciate the effort here. Building OpenCV is typically quite difficult unfortunately. The fact that we also need to upgrade this feedstock to `conda-build` 3 complicates matters. Thankfully @hajapy has already worked pretty hard on this problem. Think there is a subtlety with the macOS build failure. Have tried to elaborate on it above. Please feel free to ask questions. If we are lucky, that is the last build issue we have to deal with for this release (famous last words).

jakirkham commented on recipe/meta.yaml:
```diff
@@ -1,36 +1,39 @@
+    sha256: 4eef85759d5450b183459ff216b4c0fa43e87a4f6aa92c8af649f89336f002ec
+    patches:
+      # https://github.com/opencv/opencv/pull/12465
+      - 12465-bayer2rgba.patch
```
Normally we don't maintain patches at this feedstock as there are usually many that people want to add and the maintenance burden of adding these is quite high. Should add this is not true of all feedstocks, but has been historically true of the OpenCV feedstock in particular. IOW the builds take a while, the builds frequently fail, maintainers are on the hook for these patches as contributors don't normally maintain them, etc..

That said, if this is particularly important, we can let it slide. It would just be good not to be accumulating a bunch more patches for OpenCV going forward. Instead we should encourage them to make a new release so we can integrate all upstreamed changes as part of the release process.

hmaarrfk replied:
I get it. Ill take it off

Do you mind reviewing this pr so that I can upload the package to my own channel?

https://github.com/conda-forge/conda-forge-ci-setup-feedstock/pull/21

jakirkham commented on recipe/meta.yaml:
```diff
@@ -1,36 +1,39 @@
 
 build:
-  number: 201
+  number: 1
```
This should be set to `200`.

ref: https://conda-forge.org/docs/blas.html

hmaarrfk replied:
Ohh ok. Will do when I get home.

hmaarrfk commented:
Thanks @jakirkham that worked!

If you missed my earlier comment, please note that I’ve included a patch that was recently accepted upstream. We should be able to remove it in the 3.4.4 release.

hmaarrfk commented:
@jakirkham putting this here as the comment above will probably be lost by github when I make the change to the line.

I don't know how much pull i I will have with asking opencv for a new release.
Their last few ones have been months apart.

This is the functionality that made me look into compiling OpenCV (again) for myself.

I totally understand that you don't want to set a precedent for patches being included. 
Please consider this 1 liner PR to conda-smithy which would help me keep contributing to conda-forge:
https://github.com/conda-forge/conda-forge-ci-setup-feedstock/pull/21

I won't hold this PR hostage, i'll just wait until tomorrow to revert the patch on this one.

hmaarrfk commented on recipe/meta.yaml:
```diff
@@ -1,36 +1,38 @@
     - {{ compiler('cxx') }}
   host:
     - python
-    - pkgconfig                    # [unix]
```
I still can't figure out the difference between host and build....

jakirkham replied:
Build is where build tools are installed. Host is where dependencies that are linked to or otherwise will show up in (or influence) the run environment are added.

hmaarrfk replied:
I guess I don't know enough about build tools.

For me, there is `before ./configure; make; make install` and after `make install`

hajapy replied:
This is a conda-build 3 construct, see: https://conda.io/docs/user-guide/tasks/build-packages/define-metadata.html#requirements-section

hmaarrfk replied:
I think I get it a little more. Man I got confused between you and john because you have a very similar avatar :/

hajapy commented:
Thanks for all the hard work! This looks ready to go IMO, with the exception that you have not added yourself as a maintainer @hmaarrfk, which I'd encourage you to do :) 

I will defer merging & final review to @jakirkham.

hmaarrfk commented:
@jakirkham please don't merge just yet. One of the versions of the package is giving me really weird pinnings. I'm going to try and rebuild locally to see if that is still the case.


<details>

BTW, I'm getting some weird pinnings on this package


```
➤ conda install opencv --use-local --offline                                                                                                                                                                 (owl) 
Solving environment: done

## Package Plan ##

  environment location: /home/mark/miniconda3/envs/owl

  added / updated specs: 
    - opencv


The following packages will be UPDATED:

    blas:       1.0-mkl               --> 1.1-openblas                         conda-forge
    numpy:      1.15.1-py36h1d66e8a_0 --> 1.15.1-py36_blas_openblashd3ea46f_1  conda-forge [blas_openblas]
    opencv:     3.4.1-py36h6fd60c2_1  --> 3.4.3-py36_blas_openblash829a850_1   local       [blas_openblas]
    scipy:      1.1.0-py36hc49cb51_0  --> 1.1.0-py36_blas_openblash7943236_201 conda-forge [blas_openblas]

The following packages will be DOWNGRADED:

    numpy-base: 1.15.1-py36h81de0dd_0 --> 1.14.3-py36h0ea5e3f_1                           

Proceed ([y]/n)? y

Preparing transaction: done
Verifying transaction: done
Executing transaction: done
```

why is numpy-base being downgrade to 14.3?

And now when I try to install flake8:

```
➤ conda install flake8 --use-local                                                                                                                                                                           (owl) 
Solving environment: done

## Package Plan ##

  environment location: /home/mark/miniconda3/envs/owl

  added / updated specs: 
    - flake8


The following NEW packages will be INSTALLED:

    flake8:      3.5.0-py36_0                         conda-forge

The following packages will be UPDATED:

    numpy-base:  1.14.3-py36h0ea5e3f_1                            --> 1.15.1-py36h81de0dd_0            

The following packages will be DOWNGRADED:

    blas:        1.1-openblas                         conda-forge --> 1.0-mkl                          
    numpy:       1.15.1-py36_blas_openblashd3ea46f_1  conda-forge [blas_openblas] --> 1.15.1-py36h1d66e8a_0            
    opencv:      3.4.3-py36_blas_openblash829a850_1   local       [blas_openblas] --> 3.4.1-py36h6fd60c2_1             
    pycodestyle: 2.4.0-py_1                           conda-forge --> 2.3.1-py_1            conda-forge
    pyflakes:    2.0.0-py_0                           conda-forge --> 1.6.0-py_1            conda-forge
    scipy:       1.1.0-py36_blas_openblash7943236_201 conda-forge [blas_openblas] --> 1.1.0-py36hc49cb51_0
```
</details>

hmaarrfk commented:
so it isn't about the fact that we didn't specify a version for numpy in `requirements`?

hmaarrfk commented:
sigh, i'm convinced, but this is going to break things on one of my systems one day :/

jakirkham commented:
Thanks @hmaarrfk 😄

hmaarrfk commented:
Thank you @jakirkham those two patches were accepted upstream so hopefully they won't be a pain to maintain.
